lib/cypress/record_filter.rb

```diff
@@ -0,0 +1,94 @@
+module Cypress
+  class RecordFilter
+    def self.filter(records, filters, options)
+      records.where create_query(filters, options)
+    end
+
+    def self.create_query(input_filters, options)
+      query = {}
+
+      if input_filters['races']
+        query['race.code'] = { '$in' => input_filters['races'] }
+      end
+
+      if input_filters['ethnicities']
+        query['ethnicity.code'] = { '$in' => input_filters['ethnicities'] }
+      end
+
+      if input_filters['genders']
+        query['gender'] = { '$in' => input_filters['genders'] }
+      end
+
+      if input_filters['payers']
+        query['insurance_providers'] = { '$elemMatch' => { 'payer.name' => { '$in' => input_filters['payers'] } } }
+      end
+
+      if input_filters['age']
+        create_age_query(query, input_filters['age'], options)
+      end
+
+      if input_filters['problems']
+        create_problem_query(query, input_filters['problems'])
+      end
+
```

Comment:
You may want to reconsider query being a simple hash.  You can make this an $and query instead and add the other queries into an array.  So this would be something along the lines of this

queries = []
query = {'$and' => queries}

after this you can put all of the other queries into the array and also alleviate your worry declared below about possibly needing more than one $or query.  You would also simplify some of the other method calls by not needing to pass the query hash around.